# West the Mozzies Nile? (Part 1)

## **Contents**  

### **Part 1**
#### **Overview**
- **Background Information**
- **Problem Statement**
- **Method(ology)**

#### **Data Cleaning**
- **Train Dataset**   
- **Test Dataset**  
- **Weather Dataset**  
- **Spray Dataset**

#### **Summary**
---

<div id="overview"></div>

## **Overview**

It would be a truism to state that Covid-19 has thrown the world into a state of disarray, with 'pandemic' being the buzz-word of the decade. Doubtless, the ongoing pandemic demonstrates the importance of disease prevention, mitigation and response in preventing disruptions. At present, the disease in question is the West Nile Virus, which has been rampant in the United States since the turn of the century ([*source*](https://www.cdc.gov/westnile/statsmaps/index.html)). 


<div id="backgroundinformation"></div>

### **Background Information**

<img src = "images/wnv_us_states.png" alt = "wnv_us_states"/>

*Fig. 1 West Nile Virus Incidence by US State* ([*source*](https://www.cdc.gov/westnile/statsmaps/preliminarymapsdata2021/incidencestate-2021.html![image.png](attachment:image.png)))

As advised by the [*Centers of Disease Control and Prevention (CDC)*](https://www.cdc.gov/westnile/index.html#:~:text=West%20Nile%20virus%20(WNV)%20is,summer%20and%20continues%20through%20fall.), the West Nile Virus (hereafter "WNV") is the leading cause of mosquito-borne disease in continental US. Although birds serve as the virus host, the WNV is transmitted to humans via female mosquitoes which become carriers through biting infected bird species ([*source*](https://www.hopkinsmedicine.org/health/conditions-and-diseases/west-nile-virus)). There are no vaccines to  or medications to treat WNV, although most people who caught WNV do not fall sick and only about 1 in 5 people develop symptoms (e.g. fever, headache, bodyache,  joint pain, vomiting, diarrhea, rash). The WNV was first detected in 1937, in the West Nile District of Uganda. Hitherto, the virus has proliferated throughout multiple continents, including Africa, Asia, Australia, South America and North America ([*source*](https://www.ecdc.europa.eu/en/west-nile-fever/facts#:~:text=WNV%20was%20first%20isolated%20in,%2C%20Europe%2C%20Asia%20and%20Oceania.)). As seen in Fig. 1 above, almost the whole of US in entirety is affected by WNV, with states in the (mid)west regions (i.e. Arizona, New Mexico, Colorado, Nebraska, South Dakota, North Dakota) experiencing the highest frequencies of WNV rates. 


<img src = "images/chicago_zoning.png" alt = "chicago_zoning"/>

*Fig. 2 Chicago Broad Landuse Plan* ([*source*](https://secondcityzoning.org/))

For the intents of this study, we zoom in on Chicago, the 'windy city' located on the northeastern periphery of Illinois. Chicago is the third largest US city by population (~2.7million) and coupled with being one of the most expansive US metropolitans, Chicago has a population density of 11,783 people per square miles, ranking 9th highest in the country ([*source*](https://worldpopulationreview.com/us-cities)). With reference to the broad landuse plan in Fig 2 above, the renowned Chicago CBD with predominantly commercial uses is located at the eastern edge of the city fronting Lake Michigan; residential areas are sprawled outwards in centrifugal fashion typical of US cities, and industrial areas are concentrated about major transport routes. In 2021, there were a total of 61 reported WNV cases and 3 associated deaths in Illinois, a size proportion of which was in Chicago ([*source*](https://www.cdc.gov/westnile/statsmaps/preliminarymapsdata2021/disease-cases-state-2021.html)). Based on the [*2021 West Nile Virus Weekly Surveillance Reports*](https://www.chicago.gov/city/en/depts/cdph/supp_info/infectious/west_nile_virus_surveillancereports.html), the WNV transmission rate peaks from early-August to mid-September, which coincides with the latter stages of Chicago's mosquito season. Thus far, the battle against WNV largely revolves around preventive measures at the individual scale ([*source*](https://www.chicago.gov/city/en/depts/cdph/supp_info/infectious/preventing_west_nilevirus.html)), although seasonal targeted pesticide spraying has been inaugurated in recent years ([*source*](https://www.chicago.gov/city/en/depts/cdph/provdrs/healthy_living/news/2021/august/city-to-spray-insecticide-wednesday-to-kill-mosquitoes.html)). 

<img src = " images/chicago_climate.png" alt = "chicago_climate"/>

*Fig. 3 Chicago Climate* ([*source*](https://www.chicago.climatemps.com/))

In terms of climate, Chicago is characterized by a typically continental with cold winters, warm summers, and frequent short fluctuations in temperature, humidity, cloudiness, and wind direction ([*source*](https://www.isws.illinois.edu/statecli/general/chicago-climate-narrative.htm)). Fig. 3 above provides a brief overview of Chicago's climate across the year. Chicago is, in particular, infamous for its harsh winters. Climatic variables will be assessed in greater detail when scrutinizing the weather dataset. 

<div id="problemstatement"></div>

### **Problem Statement**

As members of the recently formed (fictitious) Disease And Treatment Agency, division of Societal Cures In Epidemiology and New Creative Engineering (DATA-SCIENCE) research team under the University of Chicago, the Chicago Municipal Government Department of Public Health has engaged us to develop a prediction model to identify the *spatial* and *temporal* elements of WNV. Specifically, the model should not only unearth areas in Chicago that are vulnerable to WNV but shed light on when cases are expected to escalate. The results will guide the state's effort in combating WNV through the deployment of pesticides. To this end, a cost-benefit analysis grounded in model predictions is to be conducted as part of business recommendations. Afterall, extensive pesticide use results in high monetary cost (i.e. taxpayer's income) and potential health hazards. 

<div id="methodology"></div>

### **Method(ology)**

#### **Import Libraries and Datasets**

In [1]:
# General modules
import numpy as np
import pandas as pd

# Mapping/Geospatial modules
from geopy.distance import geodesic

import warnings
warnings.filterwarnings('ignore')

%matplotlib inline

In [2]:
# Import datasets
train = pd.read_csv('datasets/train.csv')
test = pd.read_csv('datasets/test.csv')
weather = pd.read_csv('datasets/weather.csv')
spray = pd.read_csv('datasets/spray.csv')

##### **Model Framework**

The objective is to develop a classification model to predict the presence of WMV (using mosquitoes as proxies) in a particular location at a given time. To ensure production modelling is holistic, we not only consider a varied set of possible models as listed below, but create multiple pipelines and run accompanying gridsearches to optimize hyperparemeters: 

                           Predicted variable (y): 1 = WNV present | 0 = WNV absent

    a) Logistic Regression - Classifies a datapoint based on a discriminative algorithm by means of establishing a probabilistic boundary between binary classes (P(x) => threshold: y = 1 | P(x) < threshold: y = 0).   
    
    b) KNeighborsClassifier (KNN)- Classifies a datapoint based on the properties of "nearest neighbours" typically measured by euclidean/manhattan distance between variables across vector space.  

    c) Random Forest - Ensemble classification algorithm comprising multiple decision trees with bagging (i.e. sampling with replacement) to predict classes based on the "wisdom of crowds"; the assemblage of models (trees) operating as a community (crowd) will inform/condition all individual constituent models (tree).  
    
    d) AdaBoost - Ensemble technique which involves amalgamating "weak learners" (typically stumps) into a single "strong learner". At each iteration, observation weights are adjusted based on the degree of prediction error.  
    
    e) Gradient Boost - Likewise, an ensemble technique which iteratively builds "weak learners" (albeit small trees as opposed to stumps) to create a final "strong learner". However, the difference lies in gradient boost attempts to fit subsequent learners to the residual errors made by the previous weaker learner. 

##### **Scoring Metrics**

To ensure model performance evaluation is robust, we consider the scoring metrics below sequentially. For all metrics, we set a benchmark of 80% (i.e. >= 0.8) given that only a limited set of predictors (mainly climatic variables) are employed. 

    i) Accuracy [(TP + TN) / (TP + FP + TN + FN)] - First, we review accuracy as the most fundamental indicator, which gives the ratio of correct predictions to total predictions. However, accuracy per se is inadequate as it does not inform on model effectiveness and precision, not to mention that it can be artificially tweaked by adjusting the threshold to achieve a biased outcome. Only models which meet the established benchment of 80% will be further evaluated. 
    
    ii) ROC-AUC - Next, we examine the ROC-AUC score as the primary indicator of a model's performance. The ROC-AUC score informs on the extent to which the model minimizes false positives and false negatives effectivly, in which case a score of close to 1 is obtained. Conversely, for a weak model, the ROC-AUC score returns ~0.5, depicting the worst case scenario where the model has no discrimination capacity to distinguish between binary classes (i.e. whether a given location at a given time will have WNV or otherwise).  
    
    iii) Sensitivity [TP / (TP + FN)] - For this study, false negatives (i.e. model predicts absence of WNV when in reality, WNV is present) are relatively intolerable as it would result in excluding potentially high-risk areas for treatment on the basis that it was falsely identified as a no-/low-risk area for WNV. Hence, sensitivity will be considered too which seeks to minimize false negatives.  
    
    iv) F1-score [2 * (Precision * Recall) / (Precision + Recall)]; Precision = [TP / (TP + FP)]; Recall = [TP / (TP + FN)] - Defined as the harmonic mean of precision and recall, the former is a measure of accuracy of positive values predicted while the latter is a proxy of true positives successfully captured by the model. Putting both together, the F1-score provides a sense of confidence in a model's predicted positive values. Further, the F1-score is especially applicable for this study which involves imbalanced classes (dataset heavily skewed towards 0 rather than 1) as it takes data distribution into account. 

*TP = True Positive (Area identified with WNV activity and indeed WNV is present)*  
*TN = True Negative (Area not identified with WNV activity and indeed WNV is absent)*  
*FP = False Positive (Area identified with WNV activity but WNV is absent in reality)*  
*FN = False Negative (Area not identified with WNV activity but WNV is present in reality)*  


<div id="datacleaning"></div>

## **Data Cleaning**

We will perform data cleaning on the datasets train, test, weather and spray.

In [3]:
# Define function to view basic information of dataframe
def df_look(df):
    print('Number of rows: ', df.shape[0])
    print('Number of columns: ', df.shape[1], '\n') 
    display(df.dtypes)
    print('\nMissing values in dataset:')
    display(df.isnull().sum())
    print('\nCount of duplicated rows in dataset: ', df[df.duplicated(keep=False)].shape[0], '\n')

<div id="traindataset"></div>

### **Train Dataset**

Data cleaning of train dataset will involve:
- Checking for null values and treating for null values (if any)
- Convert column Date to date time object and create columns DayofMonth, Month and Year
- Mapping each species type to a numerical value
- Check that values in dataset are within acceptable range
- Drop unneccessary columns
- Combining the total count of mosquitoes
- Check datatypes of columns in test dataset

In [4]:
# View basic information of the train dataset
df_look(train)

Number of rows:  10506
Number of columns:  12 



Date                       object
Address                    object
Species                    object
Block                       int64
Street                     object
Trap                       object
AddressNumberAndStreet     object
Latitude                  float64
Longitude                 float64
AddressAccuracy             int64
NumMosquitos                int64
WnvPresent                  int64
dtype: object


Missing values in dataset:


Date                      0
Address                   0
Species                   0
Block                     0
Street                    0
Trap                      0
AddressNumberAndStreet    0
Latitude                  0
Longitude                 0
AddressAccuracy           0
NumMosquitos              0
WnvPresent                0
dtype: int64


Count of duplicated rows in dataset:  1062 



Train dataset has no null values.

As the percentage of rows in train dataset containing duplicates is high at 10.1% (1062 out 10506), we will choose to investigate further and not drop these duplicated rows first.

In [5]:
# Create a copy of train dataset
train1 = train.copy()

#### **Convert DateTime**

In [6]:
# Define function to convert column Date to Date time object series and add Day of the Month, Month, Year columns
def datetime_convert(df):
    df["Date"]= pd.to_datetime(df["Date"])
    
    df['DayofMonth'] = (df['Date'].dt.strftime('%d')).astype(int)
    df['Month'] = (df['Date'].dt.strftime('%m')).astype(int)
    df['Year'] = (df['Date'].dt.strftime('%Y')).astype(int)

In [7]:
# Convert values in column Date to Date time object series
datetime_convert(train1)

# Check the first 5 rows
train1.head()

Date                                            Address  \
0 2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1 2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2 2007-05-29  6200 North Mandell Avenue, Chicago, IL 60646, USA   
3 2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   
4 2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2          CULEX RESTUANS     62    N MANDELL AVE  T007   
3  CULEX PIPIENS/RESTUANS     79     W FOSTER AVE  T015   
4          CULEX RESTUANS     79     W FOSTER AVE  T015   

              AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
1  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
2   6200  N MANDELL AVE, Chicago, IL  41.994991 -87.769279                9   
3    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   
4    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   

   NumMosquitos  WnvPresent  DayofMonth  Month  Year  
0             1           0          29      5  2007  
1             1           0          29      5  2007  
2             1           0          29      5  2007  
3             1           0          29      5  2007  
4             4           0          29      5  2007

#### **Map Species to Numerical Values**

In [8]:
# Display types of species and value counts
train1['Species'].value_counts()

CULEX PIPIENS/RESTUANS    4752
CULEX RESTUANS            2740
CULEX PIPIENS             2699
CULEX TERRITANS            222
CULEX SALINARIUS            86
CULEX TARSALIS               6
CULEX ERRATICUS              1
Name: Species, dtype: int64

In [9]:
# Display value counts of Species by WnvPresent
train1.groupby('WnvPresent')['Species'].value_counts()

WnvPresent  Species               
0           CULEX PIPIENS/RESTUANS    4490
            CULEX RESTUANS            2691
            CULEX PIPIENS             2459
            CULEX TERRITANS            222
            CULEX SALINARIUS            86
            CULEX TARSALIS               6
            CULEX ERRATICUS              1
1           CULEX PIPIENS/RESTUANS     262
            CULEX PIPIENS              240
            CULEX RESTUANS              49
Name: Species, dtype: int64

We see that there are 4 species (Culex Erraticus, Culex Salinarius, Culex Tarsalis and Culex Territans) that did not carry the West Nile Virus. There are 2 species (Culex Pipiens and Culex Restuans) that carry the West Nile Virus.

In [10]:
# Create a new column Species_Encode which maps species type to a numerical value
species_dict = {
    'CULEX PIPIENS/RESTUANS': 3,
    'CULEX RESTUANS': 2,
    'CULEX PIPIENS': 1,
    'CULEX ERRATICUS': 0,
    'CULEX SALINARIUS': 0,
    'CULEX TARSALIS': 0,
    'CULEX TERRITANS': 0
}

train1['Species_Encode'] = train1['Species'].map(species_dict)

# Display value counts of Species_Encode
train1['Species_Encode'].value_counts()

3    4752
2    2740
1    2699
0     315
Name: Species_Encode, dtype: int64

#### **Aggregate NumMosquitos**



##### Check Values in Columns are Within Acceptable Range

In [11]:
# Check values in column NumMosquitos are non-negative
print(f"{len(train1[train1['NumMosquitos']<0])} values in NumMosquitos are less than 0.")

# Check values in column WnvPresent are either 0 or 1.
print("Value Counts of column WnvPresent")
display(train1['WnvPresent'].value_counts(normalize=True))

0 values in NumMosquitos are less than 0.
Value Counts of column WnvPresent


0    0.947554
1    0.052446
Name: WnvPresent, dtype: float64

All values in column NumMosquitos are within acceptable range (not negative), and all values in column WnvPresent are either 0 or 1.

In [12]:
# Dropping columns Address, Block, Street, AddressNumberAndStreet and AddressAccuracy
train1.drop(columns = ['Address', 'Block', 'Street', 'AddressNumberAndStreet', 'AddressAccuracy'], inplace = True)
train1.head()

Date                 Species  Trap   Latitude  Longitude  \
0 2007-05-29  CULEX PIPIENS/RESTUANS  T002  41.954690 -87.800991   
1 2007-05-29          CULEX RESTUANS  T002  41.954690 -87.800991   
2 2007-05-29          CULEX RESTUANS  T007  41.994991 -87.769279   
3 2007-05-29  CULEX PIPIENS/RESTUANS  T015  41.974089 -87.824812   
4 2007-05-29          CULEX RESTUANS  T015  41.974089 -87.824812   

   NumMosquitos  WnvPresent  DayofMonth  Month  Year  Species_Encode  
0             1           0          29      5  2007               3  
1             1           0          29      5  2007               2  
2             1           0          29      5  2007               2  
3             1           0          29      5  2007               3  
4             4           0          29      5  2007               2

In [13]:
# Look at the duplicated rows
train1[train1.duplicated(keep=False)]

Date                 Species  Trap   Latitude  Longitude  \
98    2007-06-26  CULEX PIPIENS/RESTUANS  T086  41.688324 -87.676709   
99    2007-06-26  CULEX PIPIENS/RESTUANS  T086  41.688324 -87.676709   
293   2007-07-11  CULEX PIPIENS/RESTUANS  T086  41.688324 -87.676709   
295   2007-07-11  CULEX PIPIENS/RESTUANS  T086  41.688324 -87.676709   
350   2007-07-11  CULEX PIPIENS/RESTUANS  T158  41.682587 -87.707973   
...          ...                     ...   ...        ...        ...   
10124 2013-09-06  CULEX PIPIENS/RESTUANS  T900  41.974689 -87.890615   
10259 2013-09-12  CULEX PIPIENS/RESTUANS  T900  41.974689 -87.890615   
10262 2013-09-12  CULEX PIPIENS/RESTUANS  T900  41.974689 -87.890615   
10391 2013-09-19           CULEX PIPIENS  T900  41.974689 -87.890615   
10392 2013-09-19           CULEX PIPIENS  T900  41.974689 -87.890615   

       NumMosquitos  WnvPresent  DayofMonth  Month  Year  Species_Encode  
98                1           0          26      6  2007               3  
99                1           0          26      6  2007               3  
293              50           0          11      7  2007               3  
295              50           0          11      7  2007               3  
350              50           0          11      7  2007               3  
...             ...         ...         ...    ...   ...             ...  
10124            50           1           6      9  2013               3  
10259            50           0          12      9  2013               3  
10262            50           0          12      9  2013               3  
10391            13           0          19      9  2013               1  
10392            13           0          19      9  2013               1  

[1062 rows x 11 columns]

In [14]:
train1[(train1.duplicated()) & (train1['NumMosquitos'] != 50)]

Date                 Species  Trap   Latitude  Longitude  \
99    2007-06-26  CULEX PIPIENS/RESTUANS  T086  41.688324 -87.676709   
512   2007-07-18          CULEX RESTUANS  T044  41.883284 -87.705085   
819   2007-08-01          CULEX RESTUANS  T086  41.688324 -87.676709   
854   2007-08-01  CULEX PIPIENS/RESTUANS  T148  42.017430 -87.687769   
857   2007-08-01           CULEX PIPIENS  T148  42.017430 -87.687769   
...          ...                     ...   ...        ...        ...   
9157  2013-07-25          CULEX RESTUANS  T900  41.974689 -87.890615   
9229  2013-08-01  CULEX PIPIENS/RESTUANS  T158  41.682587 -87.707973   
9678  2013-08-15           CULEX PIPIENS  T900  41.974689 -87.890615   
10102 2013-09-06           CULEX PIPIENS  T063  41.825610 -87.726549   
10392 2013-09-19           CULEX PIPIENS  T900  41.974689 -87.890615   

       NumMosquitos  WnvPresent  DayofMonth  Month  Year  Species_Encode  
99                1           0          26      6  2007               3  
512               1           0          18      7  2007               2  
819               1           0           1      8  2007               2  
854               2           0           1      8  2007               3  
857               1           0           1      8  2007               1  
...             ...         ...         ...    ...   ...             ...  
9157              5           0          25      7  2013               2  
9229             46           0           1      8  2013               3  
9678              3           0          15      8  2013               1  
10102             9           0           6      9  2013               1  
10392            13           0          19      9  2013               1  

[106 rows x 11 columns]

In [15]:
# Combining the total count of mosquitoes 
train1 = train1.groupby([col for col in train1.columns if col not in ['NumMosquitos']]).sum()
train1.reset_index(inplace = True)
train1

Date                 Species  Trap   Latitude  Longitude  \
0    2007-05-29           CULEX PIPIENS  T096  41.731922 -87.677512   
1    2007-05-29  CULEX PIPIENS/RESTUANS  T002  41.954690 -87.800991   
2    2007-05-29  CULEX PIPIENS/RESTUANS  T015  41.974089 -87.824812   
3    2007-05-29  CULEX PIPIENS/RESTUANS  T048  41.867108 -87.654224   
4    2007-05-29  CULEX PIPIENS/RESTUANS  T050  41.919343 -87.694259   
...         ...                     ...   ...        ...        ...   
8605 2013-09-26          CULEX RESTUANS  T082  41.803423 -87.642984   
8606 2013-09-26          CULEX RESTUANS  T102  41.750498 -87.605294   
8607 2013-09-26          CULEX RESTUANS  T209  41.740641 -87.546587   
8608 2013-09-26          CULEX RESTUANS  T220  41.963976 -87.691810   
8609 2013-09-26          CULEX RESTUANS  T221  41.659112 -87.538693   

      WnvPresent  DayofMonth  Month  Year  Species_Encode  NumMosquitos  
0              0          29      5  2007               1             1  
1              0          29      5  2007               3             1  
2              0          29      5  2007               3             1  
3              0          29      5  2007               3             1  
4              0          29      5  2007               3             1  
...          ...         ...    ...   ...             ...           ...  
8605           0          26      9  2013               2             2  
8606           0          26      9  2013               2             1  
8607           0          26      9  2013               2             1  
8608           0          26      9  2013               2             8  
8609           0          26      9  2013               2             5  

[8610 rows x 11 columns]

In [16]:
# Check datatypes of columns in train dataset are correct
train1.dtypes

Date              datetime64[ns]
Species                   object
Trap                      object
Latitude                 float64
Longitude                float64
WnvPresent                 int64
DayofMonth                 int64
Month                      int64
Year                       int64
Species_Encode             int64
NumMosquitos               int64
dtype: object

<div id="testdataset"></div>

### **Test Dataset**

Data cleaning of test dataset will involve:
- Checking for null values and treating for null values (if any)
- Mapping each species type to a numerical value
- Convert column Date to date time object and create columns DayofMonth, Month and Year
- Drop unneccessary columns

In [17]:
# View basic info of the test dataset
df_look(test)

Number of rows:  116293
Number of columns:  11 



Id                          int64
Date                       object
Address                    object
Species                    object
Block                       int64
Street                     object
Trap                       object
AddressNumberAndStreet     object
Latitude                  float64
Longitude                 float64
AddressAccuracy             int64
dtype: object


Missing values in dataset:


Id                        0
Date                      0
Address                   0
Species                   0
Block                     0
Street                    0
Trap                      0
AddressNumberAndStreet    0
Latitude                  0
Longitude                 0
AddressAccuracy           0
dtype: int64


Count of duplicated rows in dataset:  0 



In [18]:
# Create a copy of test dataset
test1 = test.copy()

# Checking the first 5 rows
test.head()

Id        Date                                            Address  \
0   1  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1   2  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2   3  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
3   4  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
4   5  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2           CULEX PIPIENS     41   N OAK PARK AVE  T002   
3        CULEX SALINARIUS     41   N OAK PARK AVE  T002   
4         CULEX TERRITANS     41   N OAK PARK AVE  T002   

              AddressNumberAndStreet  Latitude  Longitude  AddressAccuracy  
0  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
1  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
2  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
3  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
4  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9

In [19]:
display(test1['Species'].value_counts())

# Create a new column Species_Encode which maps species type to a numerical value
species_dict = {
    'CULEX PIPIENS/RESTUANS': 3,
    'CULEX RESTUANS': 2,
    'CULEX PIPIENS': 1,
    'CULEX ERRATICUS': 0,
    'CULEX SALINARIUS': 0,
    'CULEX TARSALIS': 0,
    'CULEX TERRITANS': 0,
    'UNSPECIFIED CULEX': 0
}

CULEX PIPIENS/RESTUANS    15359
CULEX RESTUANS            14670
CULEX PIPIENS             14521
CULEX SALINARIUS          14355
CULEX TERRITANS           14351
CULEX TARSALIS            14347
UNSPECIFIED CULEX         14345
CULEX ERRATICUS           14345
Name: Species, dtype: int64

In [20]:
## Applying the same conversion from train dataset to test dataset
# Convert values in column Date to Date time object series
datetime_convert(test1)

# Create a new column Species_Encode which maps species type to a numerical value
test1['Species_Encode'] = test1['Species'].map(species_dict)

# Dropping columns Address, Block, Street, AddressNumberAndStreet and AddressAccuracy
test1.drop(columns = ['Address', 'Block', 'Street', 'AddressNumberAndStreet', 'AddressAccuracy'], inplace = True)
test1.head()

Id       Date                 Species  Trap  Latitude  Longitude  \
0   1 2008-06-11  CULEX PIPIENS/RESTUANS  T002  41.95469 -87.800991   
1   2 2008-06-11          CULEX RESTUANS  T002  41.95469 -87.800991   
2   3 2008-06-11           CULEX PIPIENS  T002  41.95469 -87.800991   
3   4 2008-06-11        CULEX SALINARIUS  T002  41.95469 -87.800991   
4   5 2008-06-11         CULEX TERRITANS  T002  41.95469 -87.800991   

   DayofMonth  Month  Year  Species_Encode  
0          11      6  2008               3  
1          11      6  2008               2  
2          11      6  2008               1  
3          11      6  2008               0  
4          11      6  2008               0

No null values in test dataset.

<div id="weatherdataset"></div>

### **Weather Dataset**

Data cleaning of weather dataset will involve:
- Converting column Date to date time object and create columns DayofMonth, Month and Year
- Checking for missing values and treating missing values
- Dropping unneccessary columns
- Feature engineering for relative humidity
- Implementing rolling for weather variables
- Combining weather information from weather dataset with train and test datasets

In [21]:
# View basic info of the weather dataset
df_look(weather)

Number of rows:  2944
Number of columns:  22 



Station          int64
Date            object
Tmax             int64
Tmin             int64
Tavg            object
Depart          object
DewPoint         int64
WetBulb         object
Heat            object
Cool            object
Sunrise         object
Sunset          object
CodeSum         object
Depth           object
Water1          object
SnowFall        object
PrecipTotal     object
StnPressure     object
SeaLevel        object
ResultSpeed    float64
ResultDir        int64
AvgSpeed        object
dtype: object


Missing values in dataset:


Station        0
Date           0
Tmax           0
Tmin           0
Tavg           0
Depart         0
DewPoint       0
WetBulb        0
Heat           0
Cool           0
Sunrise        0
Sunset         0
CodeSum        0
Depth          0
Water1         0
SnowFall       0
PrecipTotal    0
StnPressure    0
SeaLevel       0
ResultSpeed    0
ResultDir      0
AvgSpeed       0
dtype: int64


Count of duplicated rows in dataset:  0 



In [22]:
# Create a copy of weather dataset
weather1 = weather.copy()

# Convert values in column Date to Date time object series
datetime_convert(weather1)

# Check the first 5 rows
weather1.head()

Station       Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1 2007-05-01    83    50   67     14        51      56    0    2   
1        2 2007-05-01    84    52   68      M        51      57    0    3   
2        1 2007-05-02    59    42   51     -3        42      47   14    0   
3        2 2007-05-02    60    43   52      M        42      47   13    0   
4        1 2007-05-03    66    46   56      2        40      48    9    0   

   ... SnowFall PrecipTotal StnPressure SeaLevel ResultSpeed ResultDir  \
0  ...      0.0        0.00       29.10    29.82         1.7        27   
1  ...        M        0.00       29.18    29.82         2.7        25   
2  ...      0.0        0.00       29.38    30.09        13.0         4   
3  ...        M        0.00       29.44    30.08        13.3         2   
4  ...      0.0        0.00       29.39    30.12        11.7         7   

  AvgSpeed DayofMonth Month  Year  
0      9.2          1     5  2007  
1      9.6          1     5  2007  
2     13.4          2     5  2007  
3     13.4          2     5  2007  
4     11.9          3     5  2007  

[5 rows x 25 columns]

#### Check for Missing Values in Weather Dataset

We will check for missing values in weather dataset. The missing values are indicated by 'M', 'T' and '-'.

In [23]:
# Display number of rows containing 'M' in each column of weather1 dataset, M = Missing
mcount = weather1.eq('M').sum()
mc_df = pd.DataFrame([mcount.values], columns = mcount.index)
mc_df.loc[:, (mc_df != 0).any()]

Tavg  Depart  WetBulb  Heat  Cool  Depth  Water1  SnowFall  PrecipTotal  \
0    11    1472        4    11    11   1472    2944      1472            2   

   StnPressure  SeaLevel  AvgSpeed  
0            4         9         3

In [24]:
# Display number of rows containing '-' in each column of weather1 dataset, '-' = unknown
nullcount = weather1.eq('-').sum()
nc_df = pd.DataFrame([nullcount.values], columns = nullcount.index)
nc_df.loc[:, (nc_df != 0).any()]

Sunrise  Sunset
0     1472    1472

In [25]:
# Display number of rows containing 'T' in each column of weather1 dataset, T = Trace amounts
weather1['PrecipTotal'] = weather1['PrecipTotal'].str.strip()
weather1['SnowFall'] = weather1['SnowFall'].str.strip()
    
Tcount = weather1.eq('T').sum()
tc_df = pd.DataFrame([Tcount.values], columns = Tcount.index)
tc_df.loc[:, (tc_df != 0).any()]

SnowFall  PrecipTotal
0        12          318

In [26]:
# Display number of empty rows
weather1['CodeSum'] = weather1['CodeSum'].str.strip()
empty_count = weather1.eq('').sum()
empty_df = pd.DataFrame([empty_count.values], columns = empty_count.index)
empty_df.loc[:, (empty_df != 0).any()]

CodeSum
0     1609

#### **Drop Selected Climatic Variables**

From above we can see that there are many columns with missing values.

As WetBulb, PrecipToTal, StnPressure, SeaLevel and AvgSpeed contain a small number of missing values and may affect the West Nile Virus Presence, we will keep these 5 columns.

Considering that there are only 2944 rows in the dataframe, we will drop the other columns with too many missing values: Depart, Depth, Water1, Snowfall and CodeSum.   

At this point in time, we will also drop Sunrise and Sunset, but we will keep them in mind if necessarily.

In [27]:
# Dropping columns 
drop_cols = ['Depart', 'Depth', 'Water1', 'SnowFall', 'CodeSum', 'Sunrise', 'Sunset']
weather1.drop(columns = drop_cols, inplace = True)

#### **Impute Missing Values for Climatic Variables**

After dropping the columns with too many missing values, we will fill in the missing values for the remaining columns.

In [28]:
# Manually input Tavg
weather1['Tavg'] = ((weather['Tmax'] + weather['Tmin'])/2).astype(float)
weather1.head()

Station       Date  Tmax  Tmin  Tavg  DewPoint WetBulb Heat Cool  \
0        1 2007-05-01    83    50  66.5        51      56    0    2   
1        2 2007-05-01    84    52  68.0        51      57    0    3   
2        1 2007-05-02    59    42  50.5        42      47   14    0   
3        2 2007-05-02    60    43  51.5        42      47   13    0   
4        1 2007-05-03    66    46  56.0        40      48    9    0   

  PrecipTotal StnPressure SeaLevel  ResultSpeed  ResultDir AvgSpeed  \
0        0.00       29.10    29.82          1.7         27      9.2   
1        0.00       29.18    29.82          2.7         25      9.6   
2        0.00       29.38    30.09         13.0          4     13.4   
3        0.00       29.44    30.08         13.3          2     13.4   
4        0.00       29.39    30.12         11.7          7     11.9   

   DayofMonth  Month  Year  
0           1      5  2007  
1           1      5  2007  
2           2      5  2007  
3           2      5  2007  
4           3      5  2007

In [29]:
# Fill in the missing values for PrecipTotal 
weather1['PrecipTotal'].replace(['T','M'], 0, inplace = True)
weather1['PrecipTotal'] = weather1['PrecipTotal'].astype(float)

In [30]:
# Fill in the missing values as median.
for col in (mc_df.loc[:, (mc_df != 0).any()]).columns:
    if col not in ['Tavg', 'Depart', 'Depth', 'Water1', 'SnowFall']:
        weather1[col].replace('M', np.nan, inplace = True)
        weather1[col].fillna(weather1[col].median(), inplace = True)
        weather1[col] = weather1[col].astype(float)

#### **Feature Engineering for Relative Humidity**

Examining all of the above columns in the dataset, we notice the relative humidity is not mentioned. From the [Preventive Pest Control](https://www.preventivepestcontrol.com/weather-affect-mosquito-activity/), we noted that mosquitoes like warm and moist weather, the relative humidity may be an important feature to take note of.

Using August–Roche–Magnus formula, we can calculate the relative humidity from temperature and dew point of the day.

In [31]:
## Creating a relative humidity column
# function to calculate pressure
def pressure_calc(temp):
    temp_c = (temp - 32) * 5 / 9 
    return np.exp((17.625 * temp_c)/(243.04 + temp_c))

# Calculating humidity
weather1['Humidity'] = pressure_calc(weather1['DewPoint']) / pressure_calc(weather1['Tavg'])

#### **Implement Rolling for Climatic Variables**

From [Centers for Disease Control and Prevention](https://www.cdc.gov/mosquitoes/about/life-cycles/culex.html), we learnt that in the life cycle of a typical Culex Species mosquitoes, it takes about 7 to 10 days for an egg to develop into an adult mosquito. Considering the time needed for rain to become stagnant water for breeding, we decided to do a rolling of 10 days to take this life cycle into consideration.

In [32]:
## Rolling for 10 days
# Separating the weather dataframe into two dataframes based on each weather station
weather_stat1 = weather1[weather1['Station'] == 1]
weather_stat2 = weather1[weather1['Station'] == 2]

# Reset the index
weather_stat1.set_index('Date', inplace = True)
weather_stat2.set_index('Date', inplace = True)

# Rolling for 10 days for each dataframe
weather_stat1 = weather_stat1.rolling(10).mean().dropna()
weather_stat1['Station'] = 1
weather_stat1.reset_index(inplace = True)

weather_stat2 = weather_stat2.rolling(10).mean().dropna()
weather_stat2['Station'] = 2
weather_stat2.reset_index(inplace = True)

In [33]:
# Combining the two separated dataframes for the two weather stations
weather_roll = pd.concat([weather_stat1, weather_stat2])
weather_roll

Date  Station  Tmax  Tmin   Tavg  DewPoint  WetBulb  Heat  Cool  \
0    2007-05-10        1  73.4  50.7  62.05      45.2     53.5   4.1   1.4   
1    2007-05-11        1  72.1  50.8  61.45      44.3     53.0   4.5   1.2   
2    2007-05-12        1  72.6  51.2  61.90      43.7     52.9   4.1   1.2   
3    2007-05-13        1  72.9  50.9  61.90      43.0     52.7   4.1   1.2   
4    2007-05-14        1  75.3  51.6  63.45      43.6     53.6   3.4   2.0   
...         ...      ...   ...   ...    ...       ...      ...   ...   ...   
1458 2014-10-27        2  63.5  46.2  54.85      42.0     48.5  10.0   0.2   
1459 2014-10-28        2  64.9  47.1  56.00      42.0     48.9   8.9   0.2   
1460 2014-10-29        2  64.1  47.5  55.80      41.4     48.7   9.1   0.2   
1461 2014-10-30        2  62.8  46.2  54.50      40.3     47.8  10.4   0.2   
1462 2014-10-31        2  62.3  44.7  53.50      39.1     46.8  11.4   0.2   

      PrecipTotal  StnPressure  SeaLevel  ResultSpeed  ResultDir  AvgSpeed  \
0           0.013       29.323    30.048         7.88       11.9      9.89   
1           0.013       29.346    30.070         8.84        9.5     10.26   
2           0.013       29.357    30.081         8.78        9.4     10.21   
3           0.013       29.367    30.093         8.27       10.1      9.83   
4           0.013       29.359    30.085         8.92       11.4     10.48   
...           ...          ...       ...          ...        ...       ...   
1458        0.004       29.351    30.022         7.21       23.3      8.71   
1459        0.005       29.330    29.999         7.85       22.5      9.30   
1460        0.005       29.333    29.997         7.82       23.3      9.30   
1461        0.005       29.351    30.017         7.92       22.4      9.04   
1462        0.009       29.354    30.022         8.99       22.2     10.14   

      DayofMonth  Month    Year  Humidity  
0            5.5    5.0  2007.0  0.552441  
1            6.5    5.0  2007.0  0.545472  
2            7.5    5.0  2007.0  0.521589  
3            8.5    5.0  2007.0  0.508340  
4            9.5    5.0  2007.0  0.493916  
...          ...    ...     ...       ...  
1458        22.5   10.0  2014.0  0.630802  
1459        23.5   10.0  2014.0  0.602760  
1460        24.5   10.0  2014.0  0.591200  
1461        25.5   10.0  2014.0  0.594738  
1462        26.5   10.0  2014.0  0.588370  

[2926 rows x 19 columns]

#### Combine Weather Information from Weather Dataset with Train and Test Datasets

Since the weather dataset by itself is of no use, we will need to incorporate the weather data with the train and test datasets.

Station 1: CHICAGO O'HARE INTERNATIONAL AIRPORT Lat: 41.995 Lon: -87.933 Elev: 662 ft. above sea level

Station 2: CHICAGO MIDWAY INTL ARPT Lat: 41.786 Lon: -87.752 Elev: 612 ft. above sea level

In [34]:
# Create a column to identify date and station
weather_roll['Date_Station'] = weather_roll['Date'].astype(str) + '_' + weather_roll['Station'].astype(str)
weather_roll

Date  Station  Tmax  Tmin   Tavg  DewPoint  WetBulb  Heat  Cool  \
0    2007-05-10        1  73.4  50.7  62.05      45.2     53.5   4.1   1.4   
1    2007-05-11        1  72.1  50.8  61.45      44.3     53.0   4.5   1.2   
2    2007-05-12        1  72.6  51.2  61.90      43.7     52.9   4.1   1.2   
3    2007-05-13        1  72.9  50.9  61.90      43.0     52.7   4.1   1.2   
4    2007-05-14        1  75.3  51.6  63.45      43.6     53.6   3.4   2.0   
...         ...      ...   ...   ...    ...       ...      ...   ...   ...   
1458 2014-10-27        2  63.5  46.2  54.85      42.0     48.5  10.0   0.2   
1459 2014-10-28        2  64.9  47.1  56.00      42.0     48.9   8.9   0.2   
1460 2014-10-29        2  64.1  47.5  55.80      41.4     48.7   9.1   0.2   
1461 2014-10-30        2  62.8  46.2  54.50      40.3     47.8  10.4   0.2   
1462 2014-10-31        2  62.3  44.7  53.50      39.1     46.8  11.4   0.2   

      PrecipTotal  StnPressure  SeaLevel  ResultSpeed  ResultDir  AvgSpeed  \
0           0.013       29.323    30.048         7.88       11.9      9.89   
1           0.013       29.346    30.070         8.84        9.5     10.26   
2           0.013       29.357    30.081         8.78        9.4     10.21   
3           0.013       29.367    30.093         8.27       10.1      9.83   
4           0.013       29.359    30.085         8.92       11.4     10.48   
...           ...          ...       ...          ...        ...       ...   
1458        0.004       29.351    30.022         7.21       23.3      8.71   
1459        0.005       29.330    29.999         7.85       22.5      9.30   
1460        0.005       29.333    29.997         7.82       23.3      9.30   
1461        0.005       29.351    30.017         7.92       22.4      9.04   
1462        0.009       29.354    30.022         8.99       22.2     10.14   

      DayofMonth  Month    Year  Humidity  Date_Station  
0            5.5    5.0  2007.0  0.552441  2007-05-10_1  
1            6.5    5.0  2007.0  0.545472  2007-05-11_1  
2            7.5    5.0  2007.0  0.521589  2007-05-12_1  
3            8.5    5.0  2007.0  0.508340  2007-05-13_1  
4            9.5    5.0  2007.0  0.493916  2007-05-14_1  
...          ...    ...     ...       ...           ...  
1458        22.5   10.0  2014.0  0.630802  2014-10-27_2  
1459        23.5   10.0  2014.0  0.602760  2014-10-28_2  
1460        24.5   10.0  2014.0  0.591200  2014-10-29_2  
1461        25.5   10.0  2014.0  0.594738  2014-10-30_2  
1462        26.5   10.0  2014.0  0.588370  2014-10-31_2  

[2926 rows x 20 columns]

In [35]:
# Loading the lat-long data for the two weather stations
station1 = (41.995, -87.933)
station2 = (41.786, -87.752)
  
# Function to map the Closest Station for merging
def closest_station(df):
    closest_stat = []
    for ind in df.index:
        dist1 = geodesic(station1, (df['Latitude'].loc[ind], df['Longitude'].loc[ind])).km
        dist2 = geodesic(station2, (df['Latitude'].loc[ind], df['Longitude'].loc[ind])).km
        if dist1 < dist2:
            closest_stat.append('1')
        else: 
            closest_stat.append('2')
    df['Closest_Station'] = closest_stat
    df['Date_Station'] = df['Date'].astype(str) + '_' + df['Closest_Station'].astype(str)
    df.drop(columns = ['Closest_Station'], inplace = True)

In [36]:
# Mapping to the train dataset
closest_station(train1)
train1.head()

Date                 Species  Trap   Latitude  Longitude  WnvPresent  \
0 2007-05-29           CULEX PIPIENS  T096  41.731922 -87.677512           0   
1 2007-05-29  CULEX PIPIENS/RESTUANS  T002  41.954690 -87.800991           0   
2 2007-05-29  CULEX PIPIENS/RESTUANS  T015  41.974089 -87.824812           0   
3 2007-05-29  CULEX PIPIENS/RESTUANS  T048  41.867108 -87.654224           0   
4 2007-05-29  CULEX PIPIENS/RESTUANS  T050  41.919343 -87.694259           0   

   DayofMonth  Month  Year  Species_Encode  NumMosquitos  Date_Station  
0          29      5  2007               1             1  2007-05-29_2  
1          29      5  2007               3             1  2007-05-29_1  
2          29      5  2007               3             1  2007-05-29_1  
3          29      5  2007               3             1  2007-05-29_2  
4          29      5  2007               3             1  2007-05-29_2

In [37]:
# Adding the weather data to the train dataset
combined_train = train1.merge(weather_roll.drop(columns = ['Date', 'DayofMonth', 'Month', 'Year']), on = 'Date_Station')
combined_train.head()

Date                 Species  Trap   Latitude  Longitude  WnvPresent  \
0 2007-05-29           CULEX PIPIENS  T096  41.731922 -87.677512           0   
1 2007-05-29  CULEX PIPIENS/RESTUANS  T048  41.867108 -87.654224           0   
2 2007-05-29  CULEX PIPIENS/RESTUANS  T050  41.919343 -87.694259           0   
3 2007-05-29  CULEX PIPIENS/RESTUANS  T054  41.921965 -87.632085           0   
4 2007-05-29  CULEX PIPIENS/RESTUANS  T086  41.688324 -87.676709           0   

   DayofMonth  Month  Year  Species_Encode  ...  WetBulb Heat  Cool  \
0          29      5  2007               1  ...     57.4  1.8   4.8   
1          29      5  2007               3  ...     57.4  1.8   4.8   
2          29      5  2007               3  ...     57.4  1.8   4.8   
3          29      5  2007               3  ...     57.4  1.8   4.8   
4          29      5  2007               3  ...     57.4  1.8   4.8   

   PrecipTotal  StnPressure  SeaLevel  ResultSpeed  ResultDir  AvgSpeed  \
0        0.056       29.451    30.102         7.54       13.5      9.23   
1        0.056       29.451    30.102         7.54       13.5      9.23   
2        0.056       29.451    30.102         7.54       13.5      9.23   
3        0.056       29.451    30.102         7.54       13.5      9.23   
4        0.056       29.451    30.102         7.54       13.5      9.23   

   Humidity  
0   0.53668  
1   0.53668  
2   0.53668  
3   0.53668  
4   0.53668  

[5 rows x 27 columns]

In [38]:
## Repeat process for the test dataset
closest_station(test1)
combined_test = test1.merge(weather_roll.drop(columns = ['Date', 'DayofMonth', 'Month', 'Year']), on = 'Date_Station')
combined_test.head()

Id       Date                 Species  Trap  Latitude  Longitude  \
0   1 2008-06-11  CULEX PIPIENS/RESTUANS  T002  41.95469 -87.800991   
1   2 2008-06-11          CULEX RESTUANS  T002  41.95469 -87.800991   
2   3 2008-06-11           CULEX PIPIENS  T002  41.95469 -87.800991   
3   4 2008-06-11        CULEX SALINARIUS  T002  41.95469 -87.800991   
4   5 2008-06-11         CULEX TERRITANS  T002  41.95469 -87.800991   

   DayofMonth  Month  Year  Species_Encode  ... WetBulb  Heat  Cool  \
0          11      6  2008               3  ...    65.5   0.3   7.1   
1          11      6  2008               2  ...    65.5   0.3   7.1   
2          11      6  2008               1  ...    65.5   0.3   7.1   
3          11      6  2008               0  ...    65.5   0.3   7.1   
4          11      6  2008               0  ...    65.5   0.3   7.1   

   PrecipTotal  StnPressure  SeaLevel  ResultSpeed  ResultDir  AvgSpeed  \
0        0.265       29.093    29.803         8.93       15.8     10.74   
1        0.265       29.093    29.803         8.93       15.8     10.74   
2        0.265       29.093    29.803         8.93       15.8     10.74   
3        0.265       29.093    29.803         8.93       15.8     10.74   
4        0.265       29.093    29.803         8.93       15.8     10.74   

   Humidity  
0  0.719088  
1  0.719088  
2  0.719088  
3  0.719088  
4  0.719088  

[5 rows x 26 columns]

<div id="spraydataset"></div>

### **Spray Dataset**

Data cleaning of spray dataset will involve:
- Checking for missing values and treating missing values
- Removing duplicates
- Drop unneccessary column Time

In [39]:
# Looking at the basic info of the weather dataset
df_look(spray)

# Look at the first 5 rows
spray.head()

Number of rows:  14835
Number of columns:  4 



Date          object
Time          object
Latitude     float64
Longitude    float64
dtype: object


Missing values in dataset:


Date           0
Time         584
Latitude       0
Longitude      0
dtype: int64


Count of duplicated rows in dataset:  543 



Date        Time   Latitude  Longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [40]:
spray['Date'].value_counts()

2013-08-15    2668
2013-08-29    2302
2013-07-17    2202
2011-09-07    2114
2013-07-25    1607
2013-08-22    1587
2013-08-08    1195
2013-09-05     924
2013-08-16     141
2011-08-29      95
Name: Date, dtype: int64

In [41]:
## Create a copy of dataset spray 
spray1 = spray.copy()

#### **Drop Duplicate Rows and Time Column**

In [42]:
# Checking out the duplicated rows
spray1[spray1.duplicated(keep=False)].value_counts()

Date        Time        Latitude   Longitude 
2011-09-07  7:44:32 PM  41.986460  -87.794225    541
            7:43:40 PM  41.983917  -87.793088      2
dtype: int64

We see that that there are 541 rows with same values in Date (2011-09-07), Time (7:44:32 PM), Latitude (41.986460) and Longitude (-87.794225). There are 2 rows with the same values in Date (2011-09-07), Time (7:43:40 PM), Latitude (41.983917), Longitude (-87.793088). We will drop these duplicates.

In [43]:
# Drop duplicated rows and keep first occurrences
spray1 = spray1.drop_duplicates(keep='first')

# Display dimensions of dataset spray1
spray1.shape

(14294, 4)

In [44]:
# Drop column Time since Time does not exist in the other dataframes
spray1.drop(columns = 'Time', inplace = True)

# Display dimensions of dataset spray1
print(f"Dimensions of dataset spray1: {spray1.shape}")

Dimensions of dataset spray1: (14294, 3)


<div id="p1sum"></div>

## **Summary**

At this point in time, we have performed data cleaning on the datasets provided. In the aspect of feature engineering, we have mapped the species of mosquitoes to integers for the train and test datasets. We also created a new 'relative humidity' column, as well as rolled the weather data for 10 days.

Lastly, we combined the weather data by adding them to the train and test datasets. We will now save all these new dataframes to csv files in order to prep them for analysis.

### Saving the dataframes to csv

In [45]:
# Renaming the columns of all the dataframes to lowercase
combined_train.columns = combined_train.columns.str.lower()
combined_test.columns = combined_test.columns.str.lower()
weather_roll.columns = weather_roll.columns.str.lower()
spray1.columns = spray1.columns.str.lower()

In [46]:
combined_train.to_csv('datasets/train_clean.csv', index = False)
combined_test.to_csv('datasets/test_clean.csv', index = False)
weather_roll.to_csv('datasets/weather_roll.csv', index = False)
spray1.to_csv('datasets/spray_clean.csv', index = False)